pyfr/integrators/writers.py

```diff
@@ -76,28 +76,90 @@ def __init__(self, *args, **kwargs):
-                prank = self.rallocs.mprankmap[mrank]
-                for tag, (etype, dims) in enumerate(meleinfo):
-                    name = self._get_name_for_soln(etype, prank)
-
-                    if mrank == root:
-                        loc_names.append(name)
-                    else:
-                        rbuf = np.empty(dims, dtype=self.backend.fpdtype)
-                        rreq = comm.Recv_init(rbuf, mrank, tag)
-
-                        mpi_rbufs.append(rbuf)
-                        mpi_rreqs.append(rreq)
-                        mpi_names.append(name)
+        # Deciding if parallel
+        parallel = h5py.get_config().mpi
+        parallel &= h5py.version.version_tuple[:2] >= (2, 5)
+        parallel &= not self.cfg.getbool('soln-output', 'serial-h5', False)
+
+        if parallel:
+            self._write = self._writeParallel
+        else:
+            self._write = self._writeSerial
+
+        if parallel:
+            if rank == root:
+                solList = []
+                for mrank, meleinfo in enumerate(eleinfo):
+                    prank = self.rallocs.mprankmap[mrank]
+                    solList.extend(
+                                (self._get_name_for_soln(etype, prank), dims)
+                                for (etype, dims) in meleinfo
+                            )
+            else:
+                solList = None
+            self.solList = comm.bcast(solList, root=root)
+
+        else:
+            if rank == root:
+                self._mpi_rbufs = mpi_rbufs = []
+                self._mpi_rreqs = mpi_rreqs = []
+                self._mpi_names = mpi_names = []
+                self._loc_names = loc_names = []
+
+                for mrank, meleinfo in enumerate(eleinfo):
+                    prank = self.rallocs.mprankmap[mrank]
+                    for tag, (etype, dims) in enumerate(meleinfo):
+                        name = self._get_name_for_soln(etype, prank)
+
+                        if mrank == root:
+                            loc_names.append(name)
+                        else:
+                            rbuf = np.empty(dims, dtype=self.backend.fpdtype)
+                            rreq = comm.Recv_init(rbuf, mrank, tag)
+
+                            mpi_rbufs.append(rbuf)
+                            mpi_rreqs.append(rreq)
+                            mpi_names.append(name)
 
     def _write(self, path, solnmap, metadata):
+        raise Exception
+
+    def _writeParallel(self, path, solnmap, metadata):
+        from mpi4py import MPI
+
+        comm, rank, root = get_comm_rank_root()
+
+        h5File = h5py.File(path, 'w', driver='mpio', comm=comm)
+        sMap = {}
+        for s, shape in self.solList:
+            sMap[s] = h5File.create_dataset(s, shape, dtype=self.backend.fpdtype)
+
+        for e in solnmap:
+            s = self._get_name_for_soln(e, self.rallocs.prank)
+            sMap[s][:] = solnmap[e]
+
+        if rank == root:
+            mMap = []
```

Comment:
​Ok if you are restricting only string=>string mapping for this version
then fine!

I would have any day preferred to dump all this info into a single pickle
called _metadata,
no assumptions, just that picklable, but for the requirement that this
commit be identical to legacy files.



On Mon, Mar 9, 2015 at 2:43 PM, Freddie Witherden <notifications@github.com>
wrote:

> In pyfr/integrators/writers.py
> <https://github.com/vincentlab/PyFR/pull/39#discussion_r26040562>:
>
> > +    def _writeParallel(self, path, solnmap, metadata):
> > +        from mpi4py import MPI
> > +
> > +        comm, rank, root = get_comm_rank_root()
> > +
> > +        h5File = h5py.File(path, 'w', driver='mpio', comm=comm)
> > +        sMap = {}
> > +        for s, shape in self.solList:
> > +            sMap[s] = h5File.create_dataset(s, shape, dtype=self.backend.fpdtype)
> > +
> > +        for e in solnmap:
> > +            s = self._get_name_for_soln(e, self.rallocs.prank)
> > +            sMap[s][:] = solnmap[e]
> > +
> > +        if rank == root:
> > +            mMap = []
>
> If we decide in the future that stats should be a dict then whoever makes
> this change will also update the format. Until then we're adding code to
> cover a hypothetical.
>
> More formally: anything except strings are not just not implemented, but
> not supported as part of the current file format. Moreover, you are almost
> certain to get an exception at the moment if you pass anything string like
> as it will lack an .encode method and will not allow itself to be cast to
> a NumPy Sxx so it does not buy as anything.
>
> These checks really do not buy you anything other than extra lines of
> code. Hence why PyFR seldom bothers with them.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/vincentlab/PyFR/pull/39/files#r26040562>.
>



-- 
================================
Dr. Iyer Arvind Sundaram

The reasonable man adapts himself to the world; the unreasonable one
persists in trying to adapt the world to himself.

Therefore all progress depends on the unreasonable man

- George Bernard Shaw.